File: back/api/result.go
package api

import (
	"database/sql"
	"encoding/json"
	"net/http"
)

type ResultRequest struct {
	RoomID    string `json:"roomId"`
	Team      string `json:"team"`
	Number    int    `json:"number"`
	Guesses   []int  `json:"guesses"`
	ProductID int    `json:"productId"` // ← 単一商品のID
}

func RegisterResultAPI(db *sql.DB) {
	http.HandleFunc("/api/result", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Access-Control-Allow-Origin", "*")
		w.Header().Set("Access-Control-Allow-Headers", "Content-Type")
		if r.Method == "OPTIONS" {
			w.WriteHeader(http.StatusOK)
			return
		}

		var req ResultRequest
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			http.Error(w, "invalid request", http.StatusBadRequest)
			return
		}

		// 正解の価格をDBに問い合わせ
		var actualPrice int
		err := db.QueryRow("SELECT price FROM prices WHERE id = ?", req.ProductID).Scan(&actualPrice)
		if err != nil {
			http.Error(w, "price not found", http.StatusBadRequest)
			return
		}

		// 誤差の計算
		var totalError int
		for _, guess := range req.Guesses {
			diff := guess - actualPrice
			if diff < 0 {
				diff = -diff
			}
			totalError += diff
		}

		// 誤差の平均
		averageError := float64(totalError) / float64(len(req.Guesses))

		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]interface{}{
			"roomId":       req.RoomID,
			"averageError": averageError,
		})
	})
}

File: back/api/result.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from db import get_db

router = APIRouter()

class ResultRequest(BaseModel):
    roomId: str
    team: str
    productId: int

@router.post("/result")
def get_result(req: ResultRequest):
    db = get_db()
    # latest error
    cur = db.execute(
        "SELECT error FROM guesses WHERE room_id=? AND team=? AND product_id=? ORDER BY id DESC LIMIT 1",
        (req.roomId, req.team, req.productId)
    )
    row = cur.fetchone()
    if not row:
        raise HTTPException(404, "no guess found")
    avg_err = row["error"]
    # min error
    cur = db.execute(
        "SELECT MIN(error) as m FROM guesses WHERE room_id=? AND product_id=?",
        (req.roomId, req.productId)
    )
    m = cur.fetchone()["m"]
    # winners
    cur = db.execute(
        "SELECT team FROM guesses WHERE room_id=? AND product_id=? AND error=?",
        (req.roomId, req.productId, m)
    )
    winners = [r["team"] for r in cur.fetchall()]
    return {"averageError": avg_err, "winner": winners}